### ISSUE TITLE
Cannot pull facts over SSH with snmp_facts on ASA

### ISSUE THREAD

**Issue by jean-christophe-manciot** [ansible/ansible-modules-extras#3420]
##### ISSUE TYPE
- Bug Report
##### COMPONENT NAME

snmp_facts
##### ANSIBLE VERSION

```
ansible --version
2.3.0 (commit 20161110.abc9133)
  config file = /etc/ansible/ansible.cfg
  configured module search path = Default w/o overrides
```
##### CONFIGURATION

inventory   = ./hosts
gathering = explicit
roles_path = /home/actionmystique/Program-Files/Ubuntu/Ansible/git-Ansible/Roles/roles
private_role_vars = yes
log_path = /var/log/ansible.log
fact_caching = redis
fact_caching_timeout = 86400
retry_files_enabled = False
##### OS / ENVIRONMENT
- **Local host**: Ubuntu 16.10 4.8
- **Target nodes**: ASAv 9.5
##### SUMMARY

Running snmp_facts on this particular type of target triggers a fatal error (Permission denied), whereas I can manually login into the target node with SSH or run snmp_facts without issue on IOSv targets.
We will continue to experience issues with this module as long as the API is not consistent with other networking modules (no provider, no ssh_keyfile, no password,...). Cf. the [provider proposal](https://github.com/ansible/ansible-modules-extras/issues/3176).
##### STEPS TO REPRODUCE

```
- name: Fetching SNMPv{{ snmp.new_version }} facts from the remote node
  snmp_facts:
        community: "{{ snmp.ssh.community }}"
        host: "{{ snmp.ssh.host }}"
        version: "{{ snmp.ssh.version }}"
  register: facts
```
##### EXPECTED RESULTS

Successful snmp_facts
##### ACTUAL RESULTS

```
fatal: [ASAv2]: UNREACHABLE! => {"changed": false, "msg": "Failed to connect to the host via ssh: Permission denied (publickey,password).\r\n", "unreachable": true}
```
Despite the fact that SNMP is configured on the target with the same community:
```
sh run | inc snmp   
...
snmp-server host management 172.21.100.1 community ***** version 2c
snmp-server location California
snmp-server contact xxxxxxxxxxxxxxxx
snmp-server community *****
snmp-server enable traps syslog
...
```
Also:
```
ssh admin@172.21.100.251
Type help or '?' for a list of available commands.
ASAv1> en
Password: **********
ASAv1# sh ver

Cisco Adaptive Security Appliance Software Version 9.5(2)204 
Device Manager Version 7.5(2)
...
```

**Comment by ansibot** [comment]
@ogenstad ping, this issue is waiting for your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ogenstad** [comment]
I don't have any problems running this against a virtual ASA running 9.4(1)200. I would be surprised if it wouldn't work on 9.5.

Have you verified that you are able to reach the ASA with SNMP outside of Ansible?

``` bash
snmpwalk -v2c -c public 172.21.100.251
```

And just change it to your community.

Are you running Ansible from 172.21.100.1? Based on the above config that's the only host which will be allowed to poll the ASA.

**Comment by jean-christophe-manciot** [comment]
@ogenstad 
Yes to all questions:

```
root@samsung-ubuntu:#snmpwalk -v2c -c xxxxxxxx 172.21.100.251
SNMPv2-MIB::sysDescr.0 = STRING: Cisco Adaptive Security Appliance Version 9.5(2)204
SNMPv2-MIB::sysObjectID.0 = OID: SNMPv2-SMI::enterprises.9.1.1902
DISMAN-EVENT-MIB::sysUpTimeInstance = Timeticks: (14300) 0:02:23.00
SNMPv2-MIB::sysContact.0 = STRING: manciot.jeanchristophe@gmail.com
SNMPv2-MIB::sysName.0 = STRING: ASAv1.actionmystique.net
SNMPv2-MIB::sysLocation.0 = STRING: California
SNMPv2-MIB::sysServices.0 = INTEGER: 4
IF-MIB::ifNumber.0 = INTEGER: 11
IF-MIB::ifIndex.2 = INTEGER: 2
...
root@samsung-ubuntu:#ip add show virbr5
27: virbr5: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP group default qlen 1000
    link/ether 52:54:00:9a:a6:8c brd ff:ff:ff:ff:ff:ff
    inet 172.21.100.1/24 brd 172.21.100.255 scope global virbr5
```

And from the ASAv:

```
ASAv1# traceroute 172.21.100.1
Type escape sequence to abort.
Tracing the route to 172.21.100.1

 1  172.21.100.1 0 msec 0 msec 0 msec
```

**Comment by ogenstad** [comment]
Looking at the error where you initially reported the issue it says that it fails to connect with ssh. The snmp_facts module needs to be run in local mode. Are you using `connection: local`?

**Comment by jean-christophe-manciot** [comment]
Yes through `ansible_connection: local`

**Comment by ogenstad** [comment]
Then I don't see why it would complain about ssh. Can you show your setup with the inventory file and playbook you are using?

**Comment by jean-christophe-manciot** [comment]
How does your module get the SSH credentials or private key necessary to log into the target?
Does it assume a default path of ~/.ssh/id_rsa?
What if it is located somewhere else or with a different name?

**Comment by ogenstad** [comment]
SSH is never used. This module only collects data through SNMP. I'm guessing you aren't using `connection: local`. 

This is an example setup

Inventory file - asa-host:

```
172.29.50.5

[all:vars]
# Example if you want to use Python from a virtualenv
ansible_python_interpreter = /usr/bin/env python
```

Playbook file - asa-snmp.yml:

```
---
-  hosts: all
   connection: local
   gather_facts: false

   tasks:
     - name: SNMP
       snmp_facts:
         community: cisco
         host: "{{ inventory_hostname }}"
         version: v2c
```

Running the playbook:

```
$ ansible-playbook -i asa-host asa-snmp.yml -v
Using /Users/patrick/src/tests/ansible-asa/ansible.cfg as config file

PLAY [all] *********************************************************************

TASK [SNMP] ********************************************************************
ok: [172.29.50.5] => {"ansible_facts": {"ansible_all_ipv4_addresses": ["172.29.50.5"], "ansible_interfaces": {"10": {"adminstatus": "down", "description": "", "ifindex": "10", "mac": "000c29fedd07", "mtu": "0", "name": "GigabitEthernet0/7", "operstatus": "down", "speed": "0"}, "11": {"adminstatus": "down", "description": "", "ifindex": "11", "mac": "000c29fedd11", "mtu": "0", "name": "GigabitEthernet0/8", "operstatus": "down", "speed": "0"}, "2": {"adminstatus": "down", "description": "", "ifindex": "2", "mac": "000c29feddb7", "mtu": "0", "name": "Management0/0", "operstatus": "down", "speed": "0"}, "3": {"adminstatus": "up", "description": "", "ifindex": "3", "ipv4": [{"address": "172.29.50.5", "netmask": "255.255.255.0"}], "mac": "000c29feddc1", "mtu": "1500", "name": "MANAGEMENT", "operstatus": "up", "speed": "1500"}, "4": {"adminstatus": "down", "description": "", "ifindex": "4", "mac": "000c29feddcb", "mtu": "1500", "name": "INSIDE", "operstatus": "down", "speed": "1500"}, "5": {"adminstatus": "down", "description": "", "ifindex": "5", "mac": "000c29feddd5", "mtu": "0", "name": "GigabitEthernet0/2", "operstatus": "down", "speed": "0"}, "6": {"adminstatus": "down", "description": "", "ifindex": "6", "mac": "000c29fedddf", "mtu": "0", "name": "GigabitEthernet0/3", "operstatus": "down", "speed": "0"}, "7": {"adminstatus": "down", "description": "", "ifindex": "7", "mac": "000c29fedde9", "mtu": "0", "name": "GigabitEthernet0/4", "operstatus": "down", "speed": "0"}, "8": {"adminstatus": "down", "description": "", "ifindex": "8", "mac": "000c29feddf3", "mtu": "0", "name": "GigabitEthernet0/5", "operstatus": "down", "speed": "0"}, "9": {"adminstatus": "down", "description": "", "ifindex": "9", "mac": "000c29feddfd", "mtu": "0", "name": "GigabitEthernet0/6", "operstatus": "down", "speed": "0"}}, "ansible_syscontact": "", "ansible_sysdescr": "Cisco Adaptive Security Appliance Version 9.4(1)200", "ansible_syslocation": "", "ansible_sysname": "NS2903-ASAv-1.int.ogenstad.com", "ansible_sysobjectid": "1.3.6.1.4.1.9.1.1902", "ansible_sysuptime": "1276916700"}, "changed": false}

PLAY RECAP *********************************************************************
172.29.50.5                : ok=1    changed=0    unreachable=0    failed=0

```

Could you try with that exact setup?

**Comment by jean-christophe-manciot** [comment]
My global variable "ansible_connection" wasn't included into the role somehow.
Including it solves this.
However, that means that all information travels in clear text over the network.
It's too bad considering that Ansible offers SSH encryption by default.
Could you consider adding SSH support to your module, as requested [here](https://github.com/ansible/ansible-modules-extras/issues/3176#issuecomment-259769342), especially considering that we're dealing with a firewall?

**Comment by gundalow** [comment]
@privateip @Qalthos What do you think, is this something we can do?

**Comment by ogenstad** [comment]
The name of the module is snmp_facts and the description is: "Retrieve facts for a device using SNMP."

The module also supports SNMPv3 which would make the information encrypted.

I don't see how this would work through SSH. If the device supported Python and could run the script locally.

From my perspective this bug report should be closed given that the module works as expected.

On the other hand another issue could be created with a feature request to create a facts module for Cisco ASA using ssh (like the IOS one), but that would be a completely different module than this one.

**Comment by ansibot** [comment]
@ogenstad, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ogenstad** [comment]
@gundalow, @privateip, @Qalthos, this situation is a bit strange. I have the maintainer status of this issue which has been resolved. However I don't have any access to close the issue, but still get reminders for it.

**Comment by gundalow** [comment]
@ogenstad You can use `ansibot` to change the state of the issue as detailed in https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md

bug_resolved
